gh-82300: Add track parameter to shared memory (#110778)
python/cpython: Doc/library/multiprocessing.shared_memory.rst

copying of data.

> .. class:: SharedMemory(name=None, create=False, size=0, track=True)

Reviewer: Lets make the new `track=` a keyword only argument.

   memory block may be larger or equal to the size requested.  When attaching
   to an existing shared memory block, the ``size`` parameter is ignored.

>    *track*, when enabled, registers the shared memory block with the resource

Reviewer 2: Surely there needs to be a version changed directive to clarify that `track` was added in 3.13.

   tracker process.  This process ensures proper cleanup of shared memory
   blocks even when all other processes with access to the memory have failed
   to do so (mainly due to being killed by signals).  The resource tracker is
   overzealous in certain situations and might delete a shared memory block

>    when any process with access to the shared memory has terminated.  *track*

Reviewer 2: I'm not sure that from this sentence I can follow the scenario where it is overzealous.

Reviewer 3: Indeed, this should qualified more precisely. AFAIU, the "overzealous" part is when some process _unrelated to the current process_ is also using the shared memory.

However, the resource tracker works fine (and as intended) as long as all uses of the shared memory segment have been initiated from the process that created them (for example by passing the segment to tasks running in child processes).

Author: @pitrou Would it be fair to say that the resource tracker is overzealous in any situation other than the case where all file descriptors have been inherited (as done by multiprocessing)? I hope I understand this all correctly.

Reviewer 3: Well, it seems that pickling and unpickling a `SharedMemory` object would do the right wrt. the release tracker:
https://github.com/python/cpython/blob/e990e41823981c3993c320492f76a0872db08cf9/Lib/multiprocessing/shared_memory.py#L192-L200

And pickling is how multiprocessing passes most data to related processes (such as tasks, process arguments...) except when using the "fork" method.

Author: Thanks, that clears up quite a bit. Would it be precise to say that the issue is due to duplicate resource trackers which multiprocessing somehow manages to avoid?

Reviewer 3: IMHO, the doc should explain in which _context_ that happens, not the underlying cause.

   should be set to ``False`` if there is already another process in place
   that does the bookkeeping.  In most situations, this means that *track*

>    should be set to ``False`` when *create* is set to ``False``.

Reviewer 2: Can I get a hint at other reasons why I might want to set track=False, while create=True?

Reviewer 3: Probably when, for some reason, you created the segment yourself but want multiprocessing to own it?

Author: I couldn't think of a good reason either but found it hard to exclude the possibility that someone might have some exotic use for this.

   .. method:: close()

      Closes the file descriptor/handle to the shared memory from this

>       instance.  All instances should call ``close()`` once the instance

Reviewer 2: I know the old version also used this phrasing, but isn't it odd that "the instance must close the instance"? I'd think some **process** would be the thing that closes the SharedMemory instance.

      is no longer needed.  Depending on operating system, the underlying
      memory may or may not be freed even if all handles to it have been
      closed.  To ensure proper cleanup, use the ``unlink()`` method.

   .. method:: unlink()

      Deletes the underlying shared memory block.  This should be called only
      once per shared memory block regardless of the number of handles to it.
      After requesting its deletion, a shared memory block may or may not be

>       immediately destroyed and this behavior may differ across platforms.

Reviewer 2: ```suggestion
      immediately destroyed, and this behavior may differ across platforms.
```

>       Attempts to access data inside the shared memory block after

Reviewer 2: ```suggestion
      Therefore, attempts to access data inside the shared memory block after
```

      ``unlink()`` has been called may result in memory access errors.
      To ensure proper bookkeeping, ``unlink()`` may only be called by

>       an instance with *track* enabled.

Reviewer 2: may -> must? Or should? or ought to?

As an outsider to shared memory, the relationship between close and unlink is also not clear from these docs. Should I call both? In which order? Or are they mutually exclusive?

Reviewer 3: I have to understand I don't understand this either. Why couldn't I unlink a shared memory segment that's not tracker by the resource tracker?

Author: The idea is that in all likelihood the process that created the shared memory already has a resource tracker watching over it. If another process unlinks the memory then the resource tracker will complain quite loudly at the end of execution about the user having forgotten to clean up, only to then warn about the file not being found.

I will undo this part and add a check to the resource tracker to not complain if the user did clean up, after all.

   memory block may be larger or equal to the size requested.  When attaching
   to an existing shared memory block, the ``size`` parameter is ignored.

   *track*, when enabled, registers the shared memory block with a resource
   tracker process.  This process ensures proper cleanup of shared memory
   blocks even when all other processes with access to the memory have failed

>    to do so (mainly due to being killed by signals).  Unless a Python process
>    was created using any of the :mod:`multiprocessing` facilities (such as
>    :class:`multiprocessing.Process`), it will receive its own resource tracker
>    process when accessing shared memory with *track* enabled.  This will cause
>    the shared memory to be deleted by the resource tracker of the first
>    process that terminates.  To avoid this issue, users of :mod:`subprocess` or
>    standalone Python processes should set *track* to ``False`` when there is
>    already another process in place that does the bookkeeping.  *track* has
>    an effect only on POSIX.  Windows has its own tracking and does not use the
>    resource tracker.

Reviewer 3: These explanations feel long-winded and not very helpful at the same time. How about something like this:

```suggestion
   to do so (mainly due to being killed by signals). As long as Python processes
   are spawned from a common ancestor using :mod:`multiprocessing` facilities,
   they share a common resource tracker process, and lifetime of shared memory
   segments is handled automatically among those processes. However,
   this mechanism becomes undesirable if a shared memory segment is also
   used by unrelated processes, for example if the shared memory segment name
   is given on the command line to another program. In this case, *track*
   should be set to ``False`` and the application should find other ways to
   delete a shared memory segment that is not in use anymore.

   On Windows, *track* is ignored as the operating system transparently
   deletes shared memory when all the processes referencing it are gone.
```

Author: This new version misses some nuances of the original explanation.

You mention unrelated processes, but child processes created with subprocess are affected as well! It would be a stretch to consider them unrelated. The issue affects really anything that isn't created with multiprocessing. And given that multiprocessing already offers higher level functionality I'm still left wondering whether multiprocessing users actually use SharedMemory directly.

You also do not mention the issue of multiple resource trackers. The way I stumbled upon this issue in the first place (and I assume many others) is by getting a message like the following:

```
/usr/lib/python3.11/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked shared_memory objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
/usr/lib/python3.11/multiprocessing/resource_tracker.py:237: UserWarning: resource_tracker: '/psm_3295459d': [Errno 2] No such file or directory: '/psm_3295459d'
  warnings.warn('resource_tracker: %r: %s' % (name, e))
```

Without knowing about multiple resource trackers the message above would make me think that the resource tracker is plain bugged and not able to keep track of its own resources. I would be wondering whether I did in fact leak memory or not and whether the tracker did remove anything or not.

Reviewer 3: Hmm, that's a fair point. How about the following then:
```suggestion
   to do so (mainly due to being killed by signals). As long as Python processes
   are spawned from a common ancestor using :mod:`multiprocessing` facilities,
   they share a common resource tracker process, and lifetime of shared memory
   segments is handled automatically among those processes.
   
   However, this automatic resource cleanup can become undesirable in
   other situations, such as:

   * if a shared memory segment is used by a Python process that is not
     part of the same :mod:`multiprocessing` process tree;
   * if a shared memory segment is used by a Python process that was not
     launched using :mod:`multiprocessing`;
   * if a shared memory segment is used by a non-Python process.
   
   In such situations, it is recommended that *track* is set to ``False`` and
   that the application implement its own mechanism to handle the lifetime
   of shared memory segments.

   On Windows, *track* is ignored as the operating system transparently
   deletes shared memory when all the processes referencing it are gone.
```

Author: I've taken your first and last sentence mostly unchanged and put some of the original things in between. I think subprocess is worth mentioning once.

   .. versionchanged:: 3.13 Added *track* parameter.

   .. method:: close()

      Closes the file descriptor/handle to the shared memory from this
      instance.  :meth:`close()` should be called once access to the shared
      memory block from this instance is no longer needed.  Depending
      on operating system, the underlying memory may or may not be freed
      even if all handles to it have been closed.  To ensure proper cleanup,
      use the :meth:`unlink()` method.

   .. method:: unlink()

      Deletes the underlying shared memory block.  This should be called only

>       once per shared memory block regardless of the number of handles to it.

Reviewer 3: ```suggestion
      once per shared memory block regardless of the number of handles to it,
      even in other processes.
```

      :meth:`unlink()` and :meth:`close()` can be called in any order, but
      trying to access data inside a shared memory block after :meth:`unlink()`

>       may result in memory access errors, depending on platform.

Reviewer 3: Is it a "may"? Does Windows not do this, for example?

Author: Unlink has no effect on Windows due to its reference tracking. Should we mention that specifically here?

Reviewer 3: Ah, yes, let's do it.

   memory block may be larger or equal to the size requested.  When attaching
   to an existing shared memory block, the ``size`` parameter is ignored.

   *track*, when enabled, registers the shared memory block with a resource
   tracker process.  This process ensures proper cleanup of shared memory
   blocks even when all other processes with access to the memory have failed

>    to do so (mainly due to being killed by signals).

Reviewer: ```suggestion
   *track*, when enabled, registers the shared memory block with a resource
   tracker process on platforms where the OS does not do this automatically.
   The resource tracker ensures proper cleanup of the shared memory even
   if all other processes with access to the memory exit without doing so.
```

in this suggested edit I wanted to get a mention of platforms where the OS does this for us as well as get rid of the "(mainly due to being killed by signals)" wording as that is over specific and possibly not accurate.

The details below this all seem good as is.

   Python processes created from a common ancestor using :mod:`multiprocessing`
   facilities share a single resource tracker process, and the lifetime of
   shared memory segments is handled automatically among these processes.

>    Python processes created in any other way will receive their own own

Reviewer 3: ```suggestion
   Python processes created in any other way will receive their own
```

   resource tracker when accessing shared memory with *track* enabled.
   This will cause the shared memory to be deleted by the resource tracker
   of the first process that terminates.
   To avoid this issue, users of :mod:`subprocess` or standalone Python
   processes should set *track* to ``False`` when there is already another
   process in place that does the bookkeeping.
   *track* is ignored on Windows, which has its own tracking and
   automatically deletes shared memory when all handles to it have been closed.

   .. versionchanged:: 3.13 Added *track* parameter.

   .. method:: close()

      Closes the file descriptor/handle to the shared memory from this
      instance.  :meth:`close()` should be called once access to the shared
      memory block from this instance is no longer needed.  Depending
      on operating system, the underlying memory may or may not be freed
      even if all handles to it have been closed.  To ensure proper cleanup,
      use the :meth:`unlink()` method.

   .. method:: unlink()

>       Deletes the underlying shared memory block.  This should be called only

Reviewer: phrasing nit: I'd say "at most once" rather than "only once" as it reads less like an implication that it needs to be called at all (anyone using resource tracking likely does not _need_ to).

Author: They don't need to but they should. The resource tracker right now is quite noisy when anything has not been cleaned up, so I believe this is somewhat similar to saying that any malloc should be freed at some point even though  anyone using an operating system with decent memory management does not need to.
